src/main/java/com/williamfiset/algorithms/datastructures/balancedtree/TreapTree.java

```diff
@@ -0,0 +1,175 @@
+    public Node left, right;
+
+    public Node(T value, int priority) {
+      this.value = value;
+      this.left = this.right = null;
+      this.priority = priority;
+    }
+
+    public T getValue() {
+      return value;
+    }
+
+    public void setValue(T value) {
+      this.value = value;
+    }
+
+    @Override
+    public TreePrinter.PrintableNode getLeft() {
+      return left;
+    }
+
+    @Override
+    public TreePrinter.PrintableNode getRight() {
+      return right;
+    }
+
+    @Override
+    public String getText() {
+      return value.toString();
+    }
+  }
+  // The root node of the Treap tree.
+  public Node root;
+
+  // Tracks the number of nodes inside the tree
+  private int nodeCount = 0;
+
+  public TreapTree() {
+    random = new Random();
+  }
+
+  // returns the number of nodes in the tree
+  public int size() {
+    return this.nodeCount;
+  }
+
+  public boolean contains(T value) {
+    return contains(root, value);
+  }
+
+  private boolean contains(Node node, T value) {
+    if (node == null || value == null) return false;
+
+    int cmp = value.compareTo(node.getValue());
+
+    if (cmp < 0) return contains(node.left, value);
+    else if (cmp > 0) return contains(node.right, value);
+    else return true;
+  }
+
+  public boolean isEmpty() {
+    return (nodeCount == 0);
+  }
+
+  public boolean insert(T val, int priority) {
+    if (val == null) {
+      throw new IllegalArgumentException("TreapTree does not allow null values");
+    }
+    if (!contains(root, val)) {
+      root = insert(this.root, val, priority);
+      nodeCount++;
+      return true;
+    }
+    return false;
+  }
+
+  public boolean insert(T val) {
+    return insert(val, random.nextInt(MAX_RAND_NUM));
+  }
+
+  private Node insert(Node node, T value, int priority) {
+    if (node == null) {
+      return new Node(value, priority);
+    }
+
+    int compareResult = value.compareTo(node.value);
+
+    if (compareResult < 0) {
+      node.left = insert(node.left, value, priority);
+      if (node.left.priority > node.priority) {
+        node = rightRotation(node);
+      }
+    } else if (compareResult > 0) {
+      node.right = insert(node.right, value, priority);
+      if (node.right.priority > node.priority) {
+        node = leftRotation(node);
+      }
+    }
+    return node;
+  }
+
+  private Node leftRotation(Node node) {
+    Node newParent = node.right;
+    node.right = newParent.left;
+    newParent.left = node;
+    return newParent;
+  }
+
+  private Node rightRotation(Node node) {
+    Node newParent = node.left;
+    node.left = newParent.right;
+    newParent.right = node;
+    return newParent;
+  }
+
+  public boolean remove(T elem) {
+    if (elem == null) return false;
+    if (contains(root, elem)) {
+      root = remove(root, elem);
+      nodeCount--;
+      return true;
+    }
+    return false;
+  }
+
+  private Node remove(Node t, T x) {
+    if (t != null) {
```

Comment:
tip: if you do:

```java
if (t == null) return null;
```
You can avoid having another nested level which is easier to read